File: idea-plugin/src/main/kotlin/com/gitai/commit/ModelProvider.kt
package com.gitai.commit

interface ModelProvider {
    val id: String
    fun generate(model: String, prompt: String): String
}

data class ProviderSelection(
    val provider: ModelProvider,
    val model: String
)

interface ModelProviderFactory {
    val id: String
    val defaultModel: String
        get() = ""

    fun create(settings: GitAiSettingsStateData): ModelProvider
}

class ModelProviderRegistry(
    private val factories: List<ModelProviderFactory> = listOf(
        OllamaProviderFactory(),
        DeepSeekProviderFactory(),
        OpenAiCompatibleProviderFactory()
    )
) {
    fun select(settings: GitAiSettingsStateData): ProviderSelection {
        val factory = factories.firstOrNull { it.id == settings.providerId }
            ?: factories.first { it.id == "ollama" }
        val model = settings.model.takeIf { it.isNotBlank() } ?: factory.defaultModel
        return ProviderSelection(factory.create(settings), model)
    }
}

class OllamaProviderFactory : ModelProviderFactory {
    override val id: String = "ollama"
    override val defaultModel: String = "qwen2.5-coder:7b"

    override fun create(settings: GitAiSettingsStateData): ModelProvider =
        OllamaClient(settings.ollamaBaseUrl)
}

class OpenAiCompatibleProviderFactory : ModelProviderFactory {
    override val id: String = "openai-compatible"

    override fun create(settings: GitAiSettingsStateData): ModelProvider =
        OpenAiCompatibleClient(settings.openAiCompatibleBaseUrl, settings.openAiCompatibleApiKey)
}

File: vscode-plugin/src/providers/modelProvider.ts
/**
 * Model provider interface matching the Kotlin `ModelProvider` interface.
 */
export interface ModelProvider {
    readonly id: string;
    generate(model: string, prompt: string): Promise<string>;
    generateStream(model: string, prompt: string, onChunk: (chunk: string) => void): Promise<string>;
}

export interface ProviderSelection {
    provider: ModelProvider;
    model: string;
}

export type CommitMessageStyle = 'short' | 'detailed';

export function normalizeCommitMessageStyle(style: string | undefined): CommitMessageStyle {
    return style === 'detailed' ? 'detailed' : 'short';
}

/**
 * Provider factory interface matching the Kotlin `ModelProviderFactory` interface.
 */
export interface ModelProviderFactory {
    readonly id: string;
    readonly defaultModel: string;
    create(settings: GitAiSettings): ModelProvider;
}

/**
 * Settings model matching the Kotlin `GitAiSettingsStateData`.
 */
export interface GitAiSettings {
    providerId: string;
    model: string;
    promptStyle: string;
    messageStyle: CommitMessageStyle;
    ollamaBaseUrl: string;
    deepSeekBaseUrl: string;
    aliyunBaseUrl: string;
    miniMaxBaseUrl: string;
    kimiBaseUrl: string;
    glmBaseUrl: string;
    openAiCompatibleBaseUrl: string;
    openAiCompatibleApiKey: string;
}

/**
 * Provider base URLs (matching the Kotlin companion objects).
 */
export const PROVIDER_BASE_URLS = {
    DEEPSEEK: 'https://api.deepseek.com',
    ALIYUN: 'https://dashscope.aliyuncs.com/compatible-mode/v1',
    MINIMAX: 'https://api.minimaxi.com/v1',
    KIMI: 'https://api.moonshot.cn/v1',
    GLM: 'https://open.bigmodel.cn/api/paas/v4',
} as const;

/**
 * Provider registry matching the Kotlin `ModelProviderRegistry`.
 */
export class ModelProviderRegistry {
    private factories: ModelProviderFactory[];

    constructor(factories: ModelProviderFactory[]) {
        this.factories = factories;
    }

    select(settings: GitAiSettings): ProviderSelection {
        const factory = this.factories.find(f => f.id === settings.providerId)
            ?? this.factories.find(f => f.id === 'ollama')!;
        const model = settings.model || factory.defaultModel;
        return { provider: factory.create(settings), model };
    }

    providerIds(): string[] {
        return this.factories.map(f => f.id);
    }

    defaultModelFor(providerId: string): string {
        return this.factories.find(f => f.id === providerId)?.defaultModel ?? '';
    }
}
